File: notes/1248-Fall2024/155H/code/Loan.java
package unl.soc;

public class Loan {
	
	/**
	 * Returns the given value rounded to the nearest
	 * cent (100th spot).
	 * 
	 * @param x
	 * @return
	 */
	public static double roundToCents(double x) {
		return Math.round(x * 100.0) / 100.0;
	}

	public static void main(String args[]) {

		double principle = 10000;
		int numYears = 5;
		double annualRate = 0.05;

		double monthlyRate = annualRate / 12;
		int numTerms = numYears * 12;
		double monthlyPayment = (monthlyRate * principle) / 
				(1 - Math.pow(1 + monthlyRate, -numTerms));
		monthlyPayment = roundToCents(monthlyPayment);
				
		double balance = principle;
		for(int month=1; month<numTerms; month++) {
			double monthInterest = roundToCents(monthlyRate * balance);
			double monthPrinciple = roundToCents(monthlyPayment - monthInterest);
			double newBalance = roundToCents(balance - monthPrinciple);
			System.out.printf("%d $%.2f $%.2f $%.2f $%.2f\n", month, balance, monthInterest, monthPrinciple, newBalance);
			balance = newBalance;
		}
		
		//TODO: you may need to handle the final month separately

	}

}

File: notes/1248-Fall2024/155E/code/loan.c
/**
 * CSCE 155E - Fall 2024
 *
 * Computes a loan amortization table
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>

/**
 * Returns the given amount rounded to the nearest cent
 * (rounded to the nearest hundredth)
 */
double roundToCents(double amount);

/**
 * Computes the monthly payment of a loan with the given
 * inputs.  See <a href="https://en.wikipedia.org/wiki/Amortization_calculator">here</a>
 */
double getMonthlyPayment(double principle, double annualRate, int years);

int main(int argc, char **argv) {

  double annualRate = 0.05;
  double monthlyRate = annualRate / 12;
  double principle = 10000.0;
  double balance = principle;
  int years = 5;
  int months = years * 12;

  double monthlyPayment = getMonthlyPayment(principle, annualRate, years);
  //TODO: round to the nearest cent...
  monthlyPayment = roundToCents(monthlyPayment);

  printf("Monthly Payment: $%.2f\n", monthlyPayment);

  for(int i=1; i<months; i++) {
    double monthlyInterest = roundToCents(monthlyRate * balance);
    double monthlyPrinciple = roundToCents(monthlyPayment - monthlyInterest);
    double newBalance = balance - monthlyPrinciple;
    printf("%d $%.2f   $%.2f  $%.2f $%.2f   $\n", i, balance, monthlyInterest, monthlyPrinciple, newBalance);
    balance = balance - monthlyPrinciple;
  }
  //TODO: change this to handle the last month separately

  return 0;
}

double roundToCents(double amount) {
  return round(amount * 100.0) / 100.0;
}

double getMonthlyPayment(double principle, double annualRate, int years) {
  double monthlyRate = annualRate / 12;
  int months = years * 12;
  double monthlyPayment = (monthlyRate * principle) / (1 - pow(1 + monthlyRate, -months) );
  return monthlyPayment;
}
